File: advc_2019/golang/advc_19_01.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func get_fuel_requirement(mass int) int {
	return mass/3 - 2
}

func solve_19_01_A(lines []string) int {
	result := 0
	for _, line := range lines {
		mass, _ := strconv.Atoi(line)
		fuel := get_fuel_requirement(mass)
		result += fuel
	}
	return result
}

func solve_19_01_B(lines []string) int {
	result := 0
	for _, line := range lines {
		mass, _ := strconv.Atoi(line)

		for {
			fuel := get_fuel_requirement(mass)
			if fuel <= 0 {
				break
			}
			result += fuel
			//fmt.Println("mass ", mass, "requires fuel", fuel, "total", result)
			mass = fuel
		}
	}
	return result
}

func solve_19_01() {
	content, _ := ioutil.ReadFile("../data/input01.txt")

	lines := strings.Split(string(content), "\n")

	fmt.Println("ans1", solve_19_01_A(lines))

	fmt.Println("ans2", solve_19_01_B(lines))
}

File: advc_2019/dotnet/src/advc_19_01.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Advc2019
{
    class Problem01
    {
        public static int GetFuelSimple(int mass)
        {
            return mass / 3 - 2;
        }
        public static int Solve1(List<int> arr)
        {
            int ans = 0;
            foreach (var mass in arr)
            {
                ans += GetFuelSimple(mass);
            }

            return ans;
        }

        public static int Solve2Internal(int mass)
        {
            int fuel = GetFuelSimple(mass);

            if (fuel > 0)
            {
                fuel += Solve2Internal(fuel);
            }
            else
            {
                fuel = 0;
            }

            return fuel;
        }
        public static int Solve2(List<int> arr)
        {
            int ans = 0;
            foreach (var value in arr)
            {
                ans += Solve2Internal(value);
            }

            return ans;
        }
        public static void Start()
        {
            var textData = File.ReadAllText("data/input01.txt");
            var textArr = textData.Split(Environment.NewLine);
            var intList = textArr.Select(t => int.Parse(t)).ToList();
            int ans1 = Solve1(intList);
            int ans2 = Solve2(intList);

            Console.WriteLine($"ans = {ans1}, {ans2}");
        }

    }
}


